### ISSUE TITLE
C++ modules: error when defining `VULKAN_HPP_DISPATCH_LOADER_DYNAMIC=1`

### ISSUE THREAD

**Issue by Zingam** [KhronosGroup/Vulkan-Hpp#1649]
> Furthermore, the C++ module also pre-defines VULKAN_HPP_DEFAULT_DISPATCHER_DYNAMIC_STORAGE by default.
Therefore, if you wish to use dynamic dispatch with the Vulkan-Hpp C++ module, all that is necessary is to define the
macro 	VULKAN_HPP_DISPATCH_LOADER_DYNAMIC=1 in your project.

This is how I understand the above instruction and I compile the provided `*.cppm` with:

```cmake
target_compile_definitions("VulkanCppModule"
  PUBLIC
    # Enable dynamic dispatch
    "VULKAN_HPP_DISPATCH_LOADER_DYNAMIC=1"
)

which, causes an error. 

> [build] "C:\PROGRA~1\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.37.32822\bin\Hostx64\x64\cl.exe"  /nologo /TP -DVULKAN_HPP_DISPATCH_LOADER_DYNAMIC=1 -DVULKAN_HPP_NO_STRUCT_CONSTRUCTORS -DCMAKE_INTDIR=\"Debug\" -external:IX:\GitHub\RToolBox\build\default.ninja-multi.msvc\vcpkg_installed\x64-windows\include -external:IC:\SDKs\VulkanSDK\1.3.261.0\Include -external:W0 /DWIN32 /D_WINDOWS /EHsc /Ob0 /Od /RTC1 -std:c++latest -MDd -Zi /showIncludes @CMakeFiles\RToolBox.dir\Debug\src\Renderer\Renderer.cpp.obj.modmap /FoCMakeFiles\RToolBox.dir\Debug\src\Renderer\Renderer.cpp.obj /FdCMakeFiles\RToolBox.dir\Debug\RToolBox.pdb /FS -c X:\GitHub\RToolBox\RToolBox\src\Renderer\Renderer.cpp
[build] C:\SDKs\VulkanSDK\1.3.261.0\Include\vulkan\vulkan_handles.hpp(14713): error C2039: 'defaultDispatchLoaderDynamic': is not a member of 'vk'
[build] C:\SDKs\VulkanSDK\1.3.261.0\Include\vulkan\vulkan.cppm(20): note: see declaration of 'vk'
[build] X:\GitHub\RToolBox\RToolBox\src\Renderer\Renderer.cpp(81): note: see instantiation of default argument expression required for 'void vk::Instance::destroy<vk::DispatchLoaderDynamic>(vk::Optional<const vk::AllocationCallbacks>,const Dispatch &) noexcept const'
[build]         with
[build]         [
[build]             Dispatch=vk::DispatchLoaderDynamic
[build]         ]
[build] C:\SDKs\VulkanSDK\1.3.261.0\Include\vulkan\vulkan_handles.hpp(14713): error C2065: 'defaultDispatchLoaderDynamic': undeclared identifier
[build] ninja: build stopped: subcommand failed.

**Comment by asuessenbach** [comment]
"your project" here does not mean "your VulkanHppModule project", but the project using the VulkanHppModule.

**Comment by Zingam** [comment]
Sorry for the noob question again.

This is my CMake module:|

```cmake
include_guard()

# find Vulkan SDK
find_package("Vulkan" REQUIRED)

# set up Vulkan C++ module
add_library("VulkanCppModule")
add_library("vulkan::cppm" ALIAS "VulkanCppModule")

target_compile_definitions("VulkanCppModule"
  PUBLIC
    # Disable exceptions
    # "VULKAN_HPP_NO_EXCEPTIONS"
    # Enable designated initializers
    "VULKAN_HPP_NO_STRUCT_CONSTRUCTORS"
)
target_compile_features("VulkanCppModule"
  PUBLIC
    cxx_std_23
)
target_include_directories("VulkanCppModule"
  PRIVATE
    "${Vulkan_INCLUDE_DIR}"
)
target_link_libraries("VulkanCppModule"
  PUBLIC
    "Vulkan::Vulkan"
)

target_sources("VulkanCppModule"
  PUBLIC
    FILE_SET cxx_modules TYPE CXX_MODULES
    BASE_DIRS
      "${Vulkan_INCLUDE_DIR}"
    FILES
      "${Vulkan_INCLUDE_DIR}/vulkan/vulkan.cppm"
)
```

This is how I consume it:

```diff
add_library("${PROJECT_NAME}" STATIC)
add_library("rmm::toolbox" ALIAS "${PROJECT_NAME}")

target_compile_definitions("${PROJECT_NAME}"
  PUBLIC
+    # Enable dynamic dispatch
+    "VULKAN_HPP_DISPATCH_LOADER_DYNAMIC=1"
)

include("VulkanHPP")
find_package(cxxopts CONFIG REQUIRED)

target_link_libraries("${PROJECT_NAME}"
  PUBLIC
    "vulkan::cppm"
)

When I debug I hit:

```c++
#  else
  inline ::VULKAN_HPP_NAMESPACE::DispatchLoaderStatic & getDispatchLoaderStatic()
  {
    static ::VULKAN_HPP_NAMESPACE::DispatchLoaderStatic dls;
    return dls;
  }
#    define VULKAN_HPP_DEFAULT_DISPATCHER ::VULKAN_HPP_NAMESPACE::getDispatchLoaderStatic()
#    define VULKAN_HPP_DEFAULT_DISPATCH_LOADER_DYNAMIC_STORAGE
#  endif
#endif
```

![image](https://github.com/KhronosGroup/Vulkan-Hpp/assets/3337419/dde55a49-de20-4955-b10c-05da6521017d)

**Comment by theHamsta** [comment]
> Sorry for the noob question again. I've just started experimenting with C++ modules and I still don't grasp how macros, etc features work in that context.
> 
> This is my CMake module:|
> 
> ```cmake
> include_guard()
> 
> # find Vulkan SDK
> find_package("Vulkan" REQUIRED)
> 
> # set up Vulkan C++ module
> add_library("VulkanCppModule")
> add_library("vulkan::cppm" ALIAS "VulkanCppModule")
> 
> target_compile_definitions("VulkanCppModule"
>   PUBLIC
>     # Disable exceptions
>     # "VULKAN_HPP_NO_EXCEPTIONS"
>     # Enable designated initializers
>     "VULKAN_HPP_NO_STRUCT_CONSTRUCTORS"
> )
> target_compile_features("VulkanCppModule"
>   PUBLIC
>     cxx_std_23
> )
> target_include_directories("VulkanCppModule"
>   PRIVATE
>     "${Vulkan_INCLUDE_DIR}"
> )
> target_link_libraries("VulkanCppModule"
>   PUBLIC
>     "Vulkan::Vulkan"
> )
> 
> target_sources("VulkanCppModule"
>   PUBLIC
>     FILE_SET cxx_modules TYPE CXX_MODULES
>     BASE_DIRS
>       "${Vulkan_INCLUDE_DIR}"
>     FILES
>       "${Vulkan_INCLUDE_DIR}/vulkan/vulkan.cppm"
> )
> ```
> 
> This is how I consume it:
> 
> ```diff
> add_library("${PROJECT_NAME}" STATIC)
> add_library("rmm::toolbox" ALIAS "${PROJECT_NAME}")
> 
> target_compile_definitions("${PROJECT_NAME}"
>   PUBLIC
> +    # Enable dynamic dispatch
> +    "VULKAN_HPP_DISPATCH_LOADER_DYNAMIC=1"
> )
> 
> include("VulkanHPP")
> find_package(cxxopts CONFIG REQUIRED)
> 
> target_link_libraries("${PROJECT_NAME}"
>   PUBLIC
>     "vulkan::cppm"
> )
> ```
> 
> When I debug I hit `inline ::VULKAN_HPP_NAMESPACE::DispatchLoaderStatic & getDispatchLoaderStatic()`
> 
> ```c++
> #  else
>   inline ::VULKAN_HPP_NAMESPACE::DispatchLoaderStatic & getDispatchLoaderStatic()
>   {
>     static ::VULKAN_HPP_NAMESPACE::DispatchLoaderStatic dls;
>     return dls;
>   }
> #    define VULKAN_HPP_DEFAULT_DISPATCHER ::VULKAN_HPP_NAMESPACE::getDispatchLoaderStatic()
> #    define VULKAN_HPP_DEFAULT_DISPATCH_LOADER_DYNAMIC_STORAGE
> #  endif
> #endif
> ```
> 
> ![image](https://user-images.githubusercontent.com/3337419/263733779-dde55a49-de20-4955-b10c-05da6521017d.png)
> 
> Also defining it before the import:
> 
> ```c++
> #define VULKAN_HPP_DISPATCH_LOADER_DYNAMIC 1
> 
> import vulkan;
> ```

You must define `target_compile_definitions` for the `VulkanCppModule`, defining it in your application or before the import won't have any effect. The module is already pre-compiled (or at least its interface) before you import and consume it with your application.

None of the macros will escape `VulkanCppModule` as macros can't be imported via C++20 modules. The macros you define `VulkanCppModule` or for your application won't affect each other.

**Comment by Zingam** [comment]
> You must define `target_compile_definitions` for the `VulkanCppModule`, defining it in your application or before the import won't have any effect. The module is already pre-compiled (or at least its interface) before you import and consume it with your application.
> 
> None of the macros will escape `VulkanCppModule` as macros can't be imported via C++20 modules. The macros you define `VulkanCppModule` or for your application won't affect each other.

I understand that now. I am still not sure how to enable the dynamic dispatch. It seems to me the right place would be to add it before line 353 (like I did it in my project):

https://github.com/KhronosGroup/Vulkan-Hpp/blob/23606b20f70d33436c0e9ff4ea948eed23d7ea4e/CMakeLists.txt#L353

```diff
# Build Vulkan-Hpp as a module
if( VULKAN_HPP_ENABLE_EXPERIMENTAL_CPP20_MODULES )
	# create a target to provide VulkanHpp as C++20 module
	add_library( VulkanHppModule )
	set_target_properties( VulkanHppModule PROPERTIES
		CXX_STANDARD_REQUIRED ON
		CXX_EXTENSIONS OFF )
	target_compile_features( VulkanHppModule PUBLIC cxx_std_20 )
+       target_compile_definitions( VulkanHppModule PUBLIC VULKAN_HPP_DISPATCH_LOADER_DYNAMIC=1 )
	target_sources( VulkanHppModule
		PUBLIC
		FILE_SET vulkan_module_file BASE_DIRS ${CMAKE_SOURCE_DIR} TYPE CXX_MODULES FILES vulkan/vulkan.cppm )
	target_include_directories( VulkanHppModule PUBLIC ${CMAKE_SOURCE_DIR} )
	target_include_directories( VulkanHppModule PUBLIC "${CMAKE_SOURCE_DIR}/Vulkan-Headers/include" )
endif()
```

**Comment by asuessenbach** [comment]
@Zingam you were right, that VULKAN_HPP_DISPATCH_LOADER_DYNAMIC had to be defined on both ends, your application and the module. And there was a little more, which is fixed with #1651.
Would you please verify that?

**Comment by asuessenbach** [comment]
Closing, assuming this was resolved with #1651.
